lib/engine/game/g_1862/step/buy_train.rb

```diff
@@ -0,0 +1,116 @@
+            else
+              entity.trains.empty? || room_for_any_type?(entity)
+            end
+          end
+
+          def room_for_any_type?(entity)
+            trains = entity.trains.group_by { |t| @game.train_type(t) }
+            trains.keys.size < 3 || trains.values.any? { |v| v.size < @game.train_limit_by_type(entity) }
+          end
+
+          def room_for_type?(entity, type)
+            trains = entity.trains.group_by { |t| @game.train_type(t) }
+            (trains[type]&.size || 0) < @game.train_limit_by_type(entity)
+          end
+
+          def buyable_trains(entity)
+            depot_trains = @depot.depot_trains
+            other_trains = @depot.other_trains(entity)
+
+            depot_trains = [@depot.min_depot_train] if entity.cash < @depot.min_depot_price
+
+            other_trains.reject! { |t| entity.cash < @game.used_train_price(t) }
+            other_trains.select! { |t| room_for_type?(entity, @game.train_type(t)) }
+
+            depot_trains + other_trains
+          end
+
+          def buyable_train_variants(train, entity)
+            return [] unless buyable_trains(entity).any? { |bt| bt.variants[bt.name] }
+
+            variants = train.variants.values
+            return variants if train.owned_by_corporation?
+
+            variants.reject! { |v| /^[A-H]$/ =~ v[:name] }
+            variants.select! { |v| room_for_type?(entity, @game.train_type_by_name(v[:name])) }
+            variants
+          end
+
+          def check_spend(action)
+            warranties = action.warranties || 0
+            return unless action.entity.cash < action.price + (warranties * WARRANTY_COST)
+
+            raise GameError "#{action.entity} cannot afford warranty cost of "\
+              "#{@game.format_currency(warranties * WARRANTY_COST)}"
+          end
+
+          def buy_train_action(action)
+            entity = action.entity
+            train = action.train
+            warranties = action.warranties || 0
+            if warranties.positive?
+              if warranties > warranty_limit(train)
+                raise GameError "Can only purchase #{warranty_limit(train)} warranties for train"
+              end
+
+              entity.spend(warranties * WARRANTY_COST, @game.bank)
+              suffix = warranties > 1 ? 'ies' : 'y'
+              @log << "#{entity.name} pays #{@game.format_currency(warranties * WARRANTY_COST)} "\
+                "for #{warranties} warrant#{suffix}"
+            end
+
+            super
+
+            warranties.times { train.name = train.name + '*' }
+          end
+
+          def president_may_contribute?(_entity, _shell = nil)
+            false
+          end
+
+          def fixed_price(train)
+            @game.used_train_price(train)
+          end
+
+          def warranty_text
+            'Warranties (each)'
+          end
+
+          def warranty_max
+            /A|D/.match?(@depot.depot_trains.first.sym) ? 2 : 3
```

Comment:
why are you using regex

Reply:
Since the phases/train levels are named A-H, it seemed the easiest way to check against multiple values.

Are regexps bad? As a former perl coder, I was happy to find that Ruby supported them.

Reply:
OK, removed.